x/wasm/internal/types/proposal.go

```diff
@@ -0,0 +1,263 @@
+	ProposalTypeUpdateAdmin              = "UpdateAdmin"
+	ProposalTypeClearAdmin               = "ClearAdmin"
+)
+
+var EnabledProposalTypes = map[string]struct{}{
+	ProposalTypeStoreCode:                {},
+	ProposalTypeStoreInstantiateContract: {},
+	ProposalTypeMigrateContract:          {},
+	ProposalTypeUpdateAdmin:              {},
+	ProposalTypeClearAdmin:               {},
+}
+
+type GovProposal struct {
+	Title       string `json:"title" yaml:"title"`
+	Description string `json:"description" yaml:"description"`
+}
+
+// GetTitle returns the title of a parameter change proposal.
+func (p GovProposal) GetTitle() string { return p.Title }
+
+// GetDescription returns the description of a parameter change proposal.
+func (p GovProposal) GetDescription() string { return p.Description }
+
+// ProposalRoute returns the routing key of a parameter change proposal.
+func (p GovProposal) ProposalRoute() string { return RouterKey }
+
+// ValidateBasic validates the proposal
+func (p GovProposal) ValidateBasic() error {
+	if len(strings.TrimSpace(p.Title)) == 0 {
+		return sdkerrors.Wrap(govtypes.ErrInvalidProposalContent, "proposal title cannot be blank")
+	}
+	if len(p.Title) > govtypes.MaxTitleLength {
+		return sdkerrors.Wrapf(govtypes.ErrInvalidProposalContent, "proposal title is longer than max length of %d", govtypes.MaxTitleLength)
+	}
+
+	if len(p.Description) == 0 {
+		return sdkerrors.Wrap(govtypes.ErrInvalidProposalContent, "proposal description cannot be blank")
+	}
+	if len(p.Description) > govtypes.MaxDescriptionLength {
+		return sdkerrors.Wrapf(govtypes.ErrInvalidProposalContent, "proposal description is longer than max length of %d", govtypes.MaxDescriptionLength)
+	}
+	return nil
+}
+
+type StoreCodeProposal struct {
+	GovProposal
+	// Creator is the address that "owns" the code object
+	Creator sdk.AccAddress `json:"creator" yaml:"creator"`
+	// WASMByteCode can be raw or gzip compressed
+	WASMByteCode []byte `json:"wasm_byte_code" yaml:"wasm_byte_code"`
+	// Source is a valid absolute HTTPS URI to the contract's source code, optional
+	Source string `json:"source" yaml:"source"`
+	// Builder is a valid docker image name with tag, optional
+	Builder string `json:"builder" yaml:"builder"`
+}
+
+// ProposalType returns the type
+func (p StoreCodeProposal) ProposalType() string { return ProposalTypeStoreCode }
+
+// ValidateBasic validates the proposal
+func (p StoreCodeProposal) ValidateBasic() error {
+	if err := p.GovProposal.ValidateBasic(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// String implements the Stringer interface.
+func (p StoreCodeProposal) String() string {
+	var b strings.Builder
+
+	b.WriteString(fmt.Sprintf(`Store Code Proposal:
+  Title:       %s
+  Description: %s
+  Changes:
+`, p.Title, p.Description))
+	// todo: print all data
+	return b.String()
+}
+
+type InstantiateContractProposal struct {
+	GovProposal
+	// Creator is the address that pays the init funds
```

Comment:
Interesting... is it possible to use the "community fund" as a creator address here?

But of course, if I want to create a contract and ask governance to allow it, *I* should be paying the funds, not the chain governance.

Reply:
🤔  Maybe I miss something but there are no further constraints defined in the proposal so you can put any address there. Later in the wasm keeper the bank keeper is used and bypasses all authorization.
With the "community fund" address it problematic as the account's balance would become less than what is stored in the fund's table.
For our messages the problem does not exist as we require the creators signature

Reply:
Update: there is a list of blockedAddr setup that should be protected. In the app it contains the `maccPerms` as `ModuleAccountAddrs`

Maybe we should prevent taking tokens from any of them. WDYT?
See also https://github.com/cosmos/cosmos-sdk/blob/master/x/bank/keeper/send.go#L274

Reply:
My question as mainly on how the SDK works (which maybe neither know), and when we send from a ModuleAccount vs sending from an AccAddress that represents the account.

The link is on master, maybe it is cleaner there.. I remember Send and SendFromModuleAccount as two separate calls somewhere.

Anyway, more a random question than a change request on the code

Reply:
I have created https://github.com/CosmWasm/wasmd/issues/177 to follow up on this.